### PR TITLE
charts

### PR DESCRIPTION
**Pull request by krisma**
Added: start time distribution chart, duration distribution chart, speed chart
simulator screen shot jul 16 2016 2 08 39 am
simulator screen shot jul 16 2016 2 09 00 am
simulator screen shot jul 16 2016 2 08 58 am

### CODE REVIEW

**Comment by krisma** [comment]
I don't know why it has conflicts. I merged before checkout to new branch

**Comment by shankari** [comment]
> I don't know why it has conflicts. I merged before checkout to new branch

When did you checkout new branch? I merged both my changes for ionic deploy and Sunil's changes for the goal tab yesterday.

**Comment by shankari** [comment]
Graphs look great.
Couple of comments:
- You need a label for the y-axis in the speed graph. I think that it is m/s
- since we are short of space, you can skip entries that are below the min value and above the max value. You should retain 0 entries in the middle. So you can remove 18-23 but should retain 14.

Taking a look at the code now.

**Comment by shankari** [comment]
Do you have a label for the x axis in the start time histogram? I can't see from the screenshots.

**Comment by shankari** [comment]
I'm a little surprised to see such a wide distribution in the trip durations (`5h` versus `13h`). Not for this week, but for next week, I think that it would be cool to be able to click on those outliers show a list of (not-common) trips that match them, so that we can see whether they make sense or whether we should ignore them.

We may need to add new server-side API methods to support this. Let's talk about the best way to do this.

**Comment by shankari** [comment]
And in terms of interactivity, for the speed profile in the detail view, if I click on a section, can it highlight the appropriate range in the graph?

I'll make an issue with a list of cool visualizations and we can prioritize them wrt other changes next week.

File: www/js/common/services.js [hunk 1]
```diff
@@ -119,7 +126,107 @@ angular.module('emission.main.common.services', [])
...
+        }
+        if (address["city"]) {
+          name = name + ", " + address["city"];
+        } else if (address["town"]) {
+          name = name + ", " + address["town"];
+        } else if (address["county"]) {
+          name = name + ", " + address["county"];
+        }
+        console.log("got response, setting display name to "+name);
+        switch (mode) {
```
Comment (shankari) [inline www/js/common/services.js]: as we discussed, make this return a promise instead
#66

File: www/js/common/trip-detail.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    nv.utils.windowResize(chart1.update);
+    nv.addGraph(chart1);
+    
+    
+    var distributionDuration = $scope.trip.durations;
+    var valsDuration = [];
+    var nob = Math.floor(Math.sqrt(distributionDuration.length)); // # of bins using square-root choice
+    var mMax = Math.max(...distributionDuration);
+    var mMin = Math.min(...distributionDuration);
+    var valsDuration = d3.layout.histogram().bins(nob).range([mMin, mMax])(distributionDuration);
```
Comment (shankari) [inline www/js/common/trip-detail.js]: if the range is already [mMin, mMax] then why do we see the values from 18-23 in the screenshot?

Comment (shankari) [inline www/js/common/trip-detail.js]: ah nvm, this is the distribution histogram.

File: www/js/common/trip-detail.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                                        CommonGraph, $window) {
+    $scope.tripId = $stateParams.tripId;
+    $scope.trip = CommonGraph.data.cTripId2ObjMap[$scope.tripId];
+    var distributionStartHour = $scope.trip.common_hour_distribution;
+    var valsStartHour = [];
+    for (var i = 0; i < 24; i++) {
+      if (distributionStartHour.get(i)) {
+        valsStartHour.push({'label': i + ':00', 'value': distributionStartHour.get(i) });
+      } else {
+        valsStartHour.push({'label': i + ':00', 'value': 0 });
```
Comment (shankari) [inline www/js/common/trip-detail.js]: Ah this is why. you need to get max and min appropriately and then stop pushing here.

File: www/js/common/trip-detail.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        .call(chart1)
+        .attr('width', 350).attr('height', 1000);
+
+    nv.utils.windowResize(chart1.update);
+    nv.addGraph(chart1);
+    
+    
+    var distributionDuration = $scope.trip.durations;
+    var valsDuration = [];
+    var nob = Math.floor(Math.sqrt(distributionDuration.length)); // # of bins using square-root choice
```
Comment (shankari) [inline www/js/common/trip-detail.js]: why square root? Please document for others to understand.

File: www/js/diary/detail.js [hunk 5]
```diff
@@ -32,4 +32,44 @@ angular.module('emission.main.diary.detail',['ui-leaflet',
   $scope.tripgj = DiaryHelper.directiveForTrip($scope.trip);
   
   console.log("trip.start_place = " + JSON.stringify($scope.trip.start_place));
+
+  var data  = [];
+  var totalDistance = 0;
+  for (var s in $scope.tripgj.sections) {
+    for (var p in $scope.tripgj.sections[s].properties.distances) {
+      totalDistance += $scope.tripgj.sections[s].properties.distances[p];
```
Comment (shankari) [inline www/js/diary/detail.js]: I thought that my feedback was to have speed versus time. You get the time from the points, which are ~ 30 secs apart.

**Pull request merged**